File: apps/mobile/lib/category_checker.ts
/**
 * Local/Static Category Sanity Check (Phase 3.5 Ticket 6).
 *
 * Compares raw OCR text against a static keyword dictionary to detect
 * likely category mismatches (e.g., OCR text contains 'chocolate' when
 * category is set to 'cosmetics').
 *
 * MUST NOT:
 * - Make network calls
 * - Call an LLM or ML model
 * - Auto-correct or change the category
 * - Block submission (produces a soft yellow warning only)
 */

export const CATEGORY_KEYWORDS: Record<string, string[]> = {
  packaged_food: [
    'food', 'snack', 'biscuit', 'cookie', 'namkeen', 'chocolate', 'chips',
    'juice', 'beverage', 'sweet', 'milk', 'butter', 'cheese', 'paneer',
    'atta', 'dal', 'rice', 'wheat', 'spices', 'masala', 'edible', 'candy', 'wafer',
  ],
  cosmetics: [
    'cream', 'lotion', 'soap', 'shampoo', 'perfume', 'deodorant', 'powder',
    'makeup', 'lipstick', 'nail', 'hair', 'face', 'wash', 'moisturizer', 'serum',
    'skincare', 'dermatologically',
  ],
  drugs_pharma: [
    'tablet', 'syrup', 'capsule', 'ointment', 'gel', 'injection',
    'pharmaceutical', 'medicine', 'rx', 'schedule', 'dosage', 'mg',
  ],
  textiles: [
    'cotton', 'polyester', 'silk', 'shirt', 'pant', 'garment', 'wear',
    'fabric', 'textile', 'apparel', 'clothing', 'size', 'wool', 'nylon',
  ],
  packaged_commodity: [
    'detergent', 'cleaner', 'cement', 'paint', 'steel', 'hardware', 'battery',
    'wire', 'cable', 'pipe', 'lubricant', 'engine oil',
  ],
};

export interface CategoryCheckResult {
  isMismatch: boolean;
  warningMessage: string | null;
  detectedCategory?: string;
  matchedKeyword?: string;
}

export function checkCategoryMismatch(
  ocrText: string,
  selectedCategory: string
): CategoryCheckResult {
  if (!selectedCategory || !CATEGORY_KEYWORDS[selectedCategory]) {
    return { isMismatch: false, warningMessage: null };
  }

  const textLower = ocrText.toLowerCase();

  // 1. Check if any keyword for selected category is present
  const hasSelectedKeywords = CATEGORY_KEYWORDS[selectedCategory].some((kw) =>
    textLower.includes(kw)
  );

  if (hasSelectedKeywords) {
    return { isMismatch: false, warningMessage: null };
  }

  // 2. Check if text strongly matches a different category
  for (const [otherCategory, keywords] of Object.entries(CATEGORY_KEYWORDS)) {
    if (otherCategory === selectedCategory) continue;

    const matched = keywords.find((kw) => textLower.includes(kw));
    if (matched) {
      const selectedLabel = selectedCategory.replace('_', ' ').toUpperCase();
      const otherLabel = otherCategory.replace('_', ' ').toUpperCase();
      return {
        isMismatch: true,
        detectedCategory: otherCategory,
        matchedKeyword: matched,
        warningMessage: `Category warning: OCR text contains '${matched}' suggesting ${otherLabel}, but category is ${selectedLabel}. (The selected category remains active)`,
      };
    }
  }

  return { isMismatch: false, warningMessage: null };
}

File: apps/backend/app/services/category_checker.py
"""
Category sanity check (Gap 7).
Provides a static keyword mapping for categories and a function to detect mismatches.
"""

from __future__ import annotations

# Static keywords that strongly suggest a specific category
_CATEGORY_KEYWORDS = {
    "packaged_food": {
        "food",
        "snack",
        "biscuit",
        "namkeen",
        "chocolate",
        "chips",
        "juice",
        "beverage",
        "sweet",
        "milk",
        "butter",
        "cheese",
        "paneer",
        "atta",
        "dal",
        "rice",
        "wheat",
        "spices",
        "masala",
        "edible",
    },
    "cosmetics": {
        "cream",
        "lotion",
        "soap",
        "shampoo",
        "perfume",
        "deodorant",
        "powder",
        "makeup",
        "lipstick",
        "nail",
        "hair",
        "face",
        "wash",
        "moisturizer",
        "serum",
    },
    "drugs_pharma": {
        "tablet",
        "syrup",
        "capsule",
        "ointment",
        "gel",
        "injection",
        "pharmaceutical",
        "medicine",
        "rx",
        "schedule",
    },
    "textiles": {
        "cotton",
        "polyester",
        "silk",
        "shirt",
        "pant",
        "garment",
        "wear",
        "fabric",
        "textile",
        "apparel",
        "clothing",
    },
}


def check_category_mismatch(ocr_text: str, selected_category: str) -> tuple[bool, str | None]:
    """
    Checks if the OCR text contains strong keywords belonging to a DIFFERENT category,
    while NOT containing keywords from the selected category.

    Returns: (is_mismatch, warning_message_or_none)
    """
    if selected_category not in _CATEGORY_KEYWORDS:
        return False, None

    text_lower = ocr_text.lower()

    # Check if any keywords for the *selected* category are present
    has_selected_keywords = any(kw in text_lower for kw in _CATEGORY_KEYWORDS[selected_category])

    if has_selected_keywords:
        return False, None

    # If not, check if it strongly matches another category
    for other_category, keywords in _CATEGORY_KEYWORDS.items():
        if other_category == selected_category:
            continue

        matched_keywords = [kw for kw in keywords if kw in text_lower]
        if matched_keywords:
            # We found strong keywords for another category, and none for the selected one
            other_name = other_category.replace("_", " ").title()
            warning = f"Category mismatch detected — '{matched_keywords[0]}' suggests {other_name}. Please verify your category selection."  # noqa: E501
            return True, warning

    return False, None
